colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,336 @@
+                            # gather grad[flatten] along dp group then reshape to [H/tp, W]
+                            grad = _gather(input_=grad, dim=-1, process_group=self.data_parallel_group)
+                            grad_reshape = grad.view(-1, grad_shape[1])
+                            exp_avg_sq_row = state["exp_avg_sq_row"]  # [H/tp]
+                            exp_avg_sq_col = state["exp_avg_sq_col"]  # [W]
+                            exp_avg_sq_row.mul_(beta2t).add_(update_reshape.mean(dim=-1), alpha=(1.0 - beta2t))
+                            exp_avg_sq_col.mul_(beta2t).add_(update_reshape.mean(dim=-2), alpha=(1.0 - beta2t))
+                            # reduce col
+                            dist.all_reduce(exp_avg_sq_col, group=self.tensor_parallel_group)
+                            exp_avg_sq_col.div_(self.tensor_parallel_size)
+                            update_reshape = self._approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col)
+                            update_reshape.mul_(grad_reshape)
+                            if self.use_zero:
+                                update = _split(
+                                    input_=update_reshape.view(-1), dim=-1, process_group=self.data_parallel_group
+                                )
+                            else:
+                                update = update_reshape
+                        else:
+                            update_reshape = update.view(-1, grad_shape[1])
+                            grad_reshape = grad.view(-1, grad_shape[1])
+                            exp_avg_sq_row = state["exp_avg_sq_row"]  # [H/dp/tp]
+                            exp_avg_sq_col = state["exp_avg_sq_col"]  # [W]
+                            exp_avg_sq_row.mul_(beta2t).add_(update_reshape.mean(dim=-1), alpha=(1.0 - beta2t))
+                            exp_avg_sq_col.mul_(beta2t).add_(update_reshape.mean(dim=-2), alpha=(1.0 - beta2t))
+                            # reduce col
+                            dist.all_reduce(exp_avg_sq_col, group=self.tensor_parallel_group)
+                            exp_avg_sq_col.div_(self.tensor_parallel_size)
+                            # gather row
+                            exp_avg_sq_row_gather = _gather(
+                                input_=exp_avg_sq_row, dim=-1, process_group=self.tensor_parallel_group
+                            )
+                            sq_row_meam = exp_avg_sq_row_gather.mean(dim=-1, keepdim=True)
+                            update_reshape = self._approx_sq_grad_row_parallel(
+                                exp_avg_sq_row, exp_avg_sq_col, sq_row_meam
+                            )
+                            update_reshape.mul_(grad_reshape)
+                            if self.use_zero:
+                                update = update_reshape.view(-1)
+                            else:
+                                update = update_reshape
```

Comment:
This part is too long. Extract key logic to another methods and try to reuse them

Reply:
> This part is too long. Extract key logic to another methods and try to reuse them

If they are extracted then debug could be a disaster. Their core logic actually, first determines the two different operations resulting from zero + row paralle param, which correspond to different communication strategies;